src/BloomExe/web/ServerBase.cs

```diff
@@ -116,117 +74,122 @@ private bool StartWithExceptionHandling(out Exception error)
 			catch (Exception e)
 			{
 				error = e;
-				_listener = null;
 				return false;
 			}
 		}
 
 		private bool TryStart()
 		{
-			_listener = new HttpListener();
 			_listener.AuthenticationSchemes = AuthenticationSchemes.Anonymous;
 			_listener.Prefixes.Add(PathEndingInSlash);
 			_listener.Start();
-			//			_listener.BeginGetContext(new AsyncCallback(ListenerCallback), _listener);
 
-			_listenerThread = new Thread(HandleRequests);
 			_listenerThread.Start();
 
+			for (var i = 0; i < _workers.Length; i++)
+			{
+				_workers[i] = new Thread(WaitForRequests);
+				_workers[i].Start();
+			}
+
 			return GetIsAbleToUsePort();
 		}
 
-		private bool GetIsAbleToUsePort()
+		private void HandleRequests()
 		{
-			IsAbleToUsePort = false;
-			try
+			while (_listener.IsListening)
 			{
-				var x = new WebClientWithTimeout { Timeout = 3000 };
+				var context = _listener.BeginGetContext(ContextReady, null);
 
-				IsAbleToUsePort = ("OK" == x.DownloadString(PathEndingInSlash + "testconnection"));
-			}
-			catch (Exception e)
-			{
-				IsAbleToUsePort = false;
+				if (0 == WaitHandle.WaitAny(new[] { _stop, context.AsyncWaitHandle }))
+					return;
 			}
-			return IsAbleToUsePort;
 		}
 
-		private void HandleRequests()
+		private void ContextReady(IAsyncResult ar)
 		{
-			while (_listener.IsListening)
+			// this can happen when shutting down
+			if (!_listenerThread.IsAlive) return;
+
+			lock (_queue)
 			{
-				var context = _listener.BeginGetContext(new AsyncCallback(ListenerCallback), _listener);
-				if (0 == WaitHandle.WaitAny(new[] { _stop, context.AsyncWaitHandle }))
-					return;
+				_queue.Enqueue(_listener.EndGetContext(ar));
+				_ready.Set();
 			}
 		}
 
 		/// <summary>
-		/// TODO: Note: doing this at runtime isn't as good as doing it in the installer, because we have no way of
-		/// removing these entries on uninstall (but the installer does).
+		/// The worker threads run this function
 		/// </summary>
-		private static void AddUrlAccessControlEntry()
+		private void WaitForRequests()
```

Comment:
Again, this method does a lot more than wait for requests. It loops and processes them. Maybe something like WorkerLoop? RequestProcessorLoop?